### ISSUE TITLE
Custom sort

### ISSUE THREAD

**Issue by almas1992** [spatie/laravel-query-builder#165]
when I use custom sort like that 
```php 
QueryBuilder::for(User::class)
    ->allowedFilters('name')
    ->allowedSorts(Sort::custom('some',SomeSort::class))
    ->get()
```
the custom sort is not working , but if I use like 
```php
QueryBuilder::for(User::class)
    ->allowedSorts(Sort::custom('some',SomeSort::class))
    ->allowedFilters('name')
    ->get()
```
the custom sort is working.

Must I do `allowedSorts` before I do `allowedFilters`?

**Comment by dominikb** [comment]
Hi almas1992!

Both methods can be used in any order you like and should not interfere with each other.

I tried to replicate the described behavior in a test but was unable to reproduce it.
Have you tried using a non-custom Sort? Maybe this has to do with the query your Custom-Sort is building?

**Comment by dominikb** [comment]
I'm having a look at this.

I've already implemented a [failing test](https://github.com/dominikb/laravel-query-builder/commit/8e63873379f6f19770dd22c5ee432427d4a27318) for this scenario for everyone to play around with.

**Comment by dominikb** [comment]
## The Problem
When adding new sorts to a `QueryBuilder` object, it checks if the new sort may be a duplicate and removes  those silently. This happens in `QueryBuilder::parseSorts()`.

When determining if the new Sort is a duplicate in `filterDuplicates($sorts)` the method `getQuery()` gets called. This method was overwritten in order to enable the possibility of default sorts and for allowing all sorts by default.
If the `QueryBuilder` gets created with a `Filter` at first, there are no sorts and `addDefaultSorts()` gets called.

Now, when checking for duplicate sorts, all sorts from the query are already added as `Sort::field` and later the actual custom sort will not be added to the list of allowed sorts.

```php
// Here we use the magic getters() to access the query property 'sort'
// These expressions are equal and then by default allow every sort passed in the query
// $this->request->sorts($this->defaultSort)
// $this->request->sorts
$this->allowedSorts = collect($this->request->sorts($this->defaultSort))
    ->map(function ($sort) {
    if ($sort instanceof Sort) {
        return $sort;
    }

    return Sort::field(ltrim($sort, '-'));
});
```


### Workaround
One way of avoiding this issue is to add a `Sort` as first chained method on the `QueryBuilder` instance.

### Solutions
What we probably want is to keep default sorts within a separate collection and throw an `InvalidFilterException` when an `allowedSorts()` gets added at any later point in time.

The method for allowing default sorts should probably also be re-written. I think what was actually meant to happen is to merge the `$defaultSort` with all the supplied sort parameters.

I would be down to implement the necessary changes.

**Comment by njxqlus** [comment]
Same problem here, but adding sort to first line not work for me
```
QueryBuilder::for(Flight::class)
    ->allowedFilters([Filter::exact('approval.approved')])
    ->allowedIncludes(['approval'])
    ->allowedSorts([Sort::custom('approval.approved', UserCustomSort::class)])```

My custom sort 
```
return $query->leftJoin('approvals', 'flights.approval_id', '=', 'approvals.id')
            ->orderBy('approval.approved', $descending ? 'desc' : 'asc');```

And error
```
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'approval.approved' in 'order clause' (SQL: select * from `flights` left join `approvals` on `flights`.`approval_id` = `approvals`.`id` where exists (select * from `approvals` where `flights`.`approval_id` = `approvals`.`id` and `approvals`.`approved` = 1) order by `approval`.`approved` asc limit 15 offset 0)"```

**Comment by AlexVanderbist** [comment]
Hi, sorry I've left this for so long! I've added the failing tests to the master branch and will take a look a the proposed fix asap (probably Monday, maybe this weekend). Thanks for the effort so far!

**Comment by CsFender** [comment]
+ If I set up reverse sorting, the "include" method doesn't working for me. :(
Thanks a lot!

**Comment by dominikb** [comment]
> * If I set up reverse sorting, the "include" method doesn't working for me. :(
>   Thanks a lot!

I've quickly put a test together but was unable to reproduce the described issue. Can you provide additional information or code snippets?

Here's the test I assembled for this case:
```php
public function it_executes_the_include_regardless_of_sort_order()
{
    $model = TestModel::create(['name' => 'John Doe']);
    RelatedModel::create(['test_model_id' => $model->id, 'name' => 'Marry Poppins']);

    $unsortedQuery = QueryBuilder::for(TestModel::class, new Request([
        'include' => 'related_models',
    ]))->allowedIncludes('related_models');

    $sortedQuery = QueryBuilder::for(TestModel::class, new Request([
        'sort'    => '-name',
        'include' => 'related_models',
    ]))->allowedIncludes('related_models');

    $this->assertSame(
        $unsortedQuery->first()->relationsToArray(),
        $sortedQuery->first()->relationsToArray()
    );
}
```

**Comment by CsFender** [comment]
Hi, sorry! Our implementation was wrong. :(
Have a nice day!

**Comment by AlexVanderbist** [comment]
Cool! The main issue should be fixed as of v1.17.3. If you're still experiencing the problem, feel free to open a new issue :)
